File: ouro_sdk_python/examples/client_usage.py
"""
Client Usage Example

Demonstrates using the low-level OuroClient for direct node interaction
"""

from ouro_sdk import OuroClient, Transaction


def main():
    print("🚀 Ouroboros SDK - Client Usage Example\n")

    # Create a client to interact with Ouroboros node
    client = OuroClient("http://localhost:8001")

    # 1. Check node health
    print("🏥 Checking node health...")
    healthy = client.health_check()
    print(f"   Node status: {'✅ Healthy' if healthy else '❌ Unhealthy'}\n")

    # 2. Get mainchain balance
    print("💰 Checking mainchain balance...")
    balance = client.get_balance("ouro1alice...")
    print(f"   Address: {balance.address}")
    print(f"   Balance: {balance.balance} OURO")
    print(f"   Pending: {balance.pending} OURO\n")

    # 3. List all microchains
    print("📋 Listing all microchains...")
    microchains = client.list_microchains()
    print(f"   Found {len(microchains)} microchains:")
    for mc in microchains[:5]:
        print(f"   - {mc.id}: {mc.name} (height: {mc.block_height})")
    print()

    # 4. Get specific microchain state
    if microchains:
        mc = microchains[0]
        print("🔍 Getting microchain state...")
        state = client.get_microchain_state(mc.id)
        print(f"   ID: {state.id}")
        print(f"   Name: {state.name}")
        print(f"   Owner: {state.owner}")
        print(f"   Block Height: {state.block_height}")
        print(f"   Total Transactions: {state.tx_count}")
        if state.last_anchor_height:
            print(f"   Last Anchor: Block #{state.last_anchor_height}")
        print()

        # 5. Check microchain balance
        print("💎 Checking microchain balance...")
        mc_balance = client.get_microchain_balance(mc.id, "ouro1alice...")
        print(f"   Balance on {mc.name}: {mc_balance} tokens\n")

    # 6. Submit mainchain transaction
    print("📤 Submitting mainchain transaction...")
    tx = Transaction("ouro1alice...", "ouro1bob...", 1000)
    # In production: tx.sign('private_key')
    tx.signature = "mock_signature"  # For demo only

    try:
        tx_id = client.submit_transaction(tx.to_json())
        print(f"✅ Transaction submitted: {tx_id}\n")

        # 7. Get transaction status
        print("🔎 Checking transaction status...")
        status = client.get_transaction_status(tx_id)
        print(f"   Status: {status.value}\n")
    except Exception as e:
        print(f"❌ Transaction failed: {e}\n")

    print("🎉 Client usage example complete!")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"❌ Error: {e}")
        exit(1)

File: ouro_sdk/examples/client_usage.rs
use ouro_sdk::{OuroClient, Transaction};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🚀 Ouroboros SDK - Client Usage Example\n");

    // Create a client to interact with Ouroboros node
    let client = OuroClient::new("http://localhost:8001");

    // 1. Check node health
    println!("🏥 Checking node health...");
    let healthy = client.health_check().await?;
    println!("   Node status: {}\n", if healthy { "✅ Healthy" } else { "❌ Unhealthy" });

    // 2. Get mainchain balance
    println!("💰 Checking mainchain balance...");
    let balance = client.get_balance("ouro1alice...").await?;
    println!("   Address: {}", balance.address);
    println!("   Balance: {} OURO", balance.balance);
    println!("   Pending: {} OURO\n", balance.pending);

    // 3. List all microchains
    println!("📋 Listing all microchains...");
    let microchains = client.list_microchains().await?;
    println!("   Found {} microchains:", microchains.len());
    for mc in microchains.iter().take(5) {
        println!("   - {}: {} (height: {})", mc.id, mc.name, mc.block_height);
    }
    println!();

    // 4. Get specific microchain state
    if let Some(mc) = microchains.first() {
        println!("🔍 Getting microchain state...");
        let state = client.get_microchain_state(&mc.id).await?;
        println!("   ID: {}", state.id);
        println!("   Name: {}", state.name);
        println!("   Owner: {}", state.owner);
        println!("   Block Height: {}", state.block_height);
        println!("   Total Transactions: {}", state.tx_count);
        if let Some(anchor) = state.last_anchor_height {
            println!("   Last Anchor: Block #{}", anchor);
        }
        println!();

        // 5. Check microchain balance
        println!("💎 Checking microchain balance...");
        let mc_balance = client.get_microchain_balance(&mc.id, "ouro1alice...").await?;
        println!("   Balance on {}: {} tokens\n", mc.name, mc_balance);
    }

    // 6. Submit mainchain transaction
    println!("📤 Submitting mainchain transaction...");
    let mut tx = Transaction::new("ouro1alice...", "ouro1bob...", 1000);
    // In production: tx.sign_with_key("private_key")?;
    tx.signature = "mock_signature".to_string();  // For demo only

    match client.submit_transaction(&tx).await {
        Ok(tx_id) => println!("✅ Transaction submitted: {}\n", tx_id),
        Err(e) => println!("❌ Transaction failed: {}\n", e),
    }

    // 7. Get transaction status
    println!("🔎 Checking transaction status...");
    match client.get_transaction_status(&tx.id).await {
        Ok(status) => println!("   Status: {:?}\n", status),
        Err(e) => println!("   Error: {}\n", e),
    }

    println!("🎉 Client usage example complete!");
    Ok(())
}
